### PR TITLE
Tests for the CLI

### PR DESCRIPTION
**Pull request by wpride**
Small start with more framework than tests, but this runs through selecting a case and opening a form so should give us a base line to prevent any obvious regressions. I had to get a little bit creative to deal with assertions about the `ApplicationHost` and `XFormPlayer` input/output as well as with exiting (since in many situation the CLI would just loop indefinitely).

Could be that I missed some much more obvious way to deal with these things, let me know if so!

### CODE REVIEW

File: src/test/java/org/cli/CliTests.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        // Perform case selection
+        new CliTestRun<>("basic_app/basic.ccz",
+                "basic_app/restore.xml",
+                CaseTestReader.class,
+                "2 1 5 \n");
+    }
+
+    static abstract class CliTestReader extends BufferedReader {
+
+        private String[] args;
```
Comment (amstone326) [inline src/test/java/org/cli/CliTests.java]: `args` is the `steps` you provided, right? If so, can you name it that instead?

File: src/test/java/org/cli/CliTests.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            this.outStream = outStream;
+        }
+
+        @Override
+        public String readLine() throws IOException {
+            String output = new String(outStream.toByteArray(), StandardCharsets.UTF_8);
+            processLine(index, output);
+            String ret = args[index++];
+            outStream.reset();
+            return ret;
```
Comment (amstone326) [inline src/test/java/org/cli/CliTests.java]: I'm confused why this method returns 1 of your steps

File: src/test/java/org/cli/CliTests.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        new CliTestRun<>("basic_app/basic.ccz",
+                "basic_app/restore.xml",
+                CaseTestReader.class,
+                "2 1 5 \n");
+    }
+
+    static abstract class CliTestReader extends BufferedReader {
+
+        private String[] args;
+        private int index;
```
Comment (amstone326) [inline src/test/java/org/cli/CliTests.java]: Can you clarify what `index` represents with a more descriptive name and/or comment? It's tough to follow

File: src/translate/java/org/javarosa/engine/XFormPlayer.java [hunk 4]
```diff
@@ -103,24 +108,12 @@ public void start(String formPath) throws FileNotFoundException {
         this.start(XFormUtils.getFormFromInputStream(new FileInputStream(formPath)));
     }
 
-    public void start(String formPath, Session session) throws FileNotFoundException {
```
[the remark this answers is not in the document]
Comment (wpride) [reply src/translate/java/org/javarosa/engine/XFormPlayer.java]: yes

File: src/test/java/org/cli/CliTests.java (hunk 2 above)
Comment (wpride) [reply src/test/java/org/cli/CliTests.java]: Added a comment explaining this class

**Comment by wpride** [comment]
@amstone326 Thanks for the feedback! Updated, let me know if you want to offline about anything

**Comment by ctsims** [comment]
This is awesome, Will! Thanks for making this part of normal test running.

No feedback beyond aliza's

File: build.gradle [hunk 5]
```diff
@@ -84,6 +84,7 @@ dependencies {
       testCompile sourceSets.main.output
       testCompile sourceSets.api.output
       testCompile sourceSets.cli.output
+      testCompile sourceSets.translate.output
```
Comment (amstone326) [inline build.gradle]: Also not sure what this is for?

Comment (amstone326) [reply build.gradle]: @wpride Approving the changes but if this isn't needed then just remove it before you merge

Comment (wpride) [reply build.gradle]: This is necessary because the `translate` files (specifically `XFormPlayer`) are now under test

**Pull request merged**